packages/timeline-state-resolver/src/devices/telemetrics.ts

```diff
@@ -0,0 +1,168 @@
+	private doOnTime: DoOnTime
+
+	private socket: Socket
+	private statusCode: StatusCode = StatusCode.UNKNOWN
+	private errorMessage: string
+
+	private recentlySendCalls: Map<string, number> = new Map()
+	private retryConnectionTimer: Timer | undefined
+
+	constructor(deviceId: string, deviceOptions: DeviceOptionsTelemetrics, getCurrentTime: () => Promise<number>) {
+		super(deviceId, deviceOptions, getCurrentTime)
+
+		this.doOnTime = new DoOnTime(() => this.getCurrentTime())
+		this.handleDoOnTime(this.doOnTime, 'telemetrics')
+	}
+
+	get canConnect(): boolean {
+		return true
+	}
+
+	clearFuture(_clearAfterTime: number): void {
+		// No state to handle - we use a fire and forget approach
+	}
+
+	get connected(): boolean {
+		return this.statusCode === StatusCode.GOOD
+	}
+
+	get deviceName(): string {
+		return `${TELEMETRICS_NAME} ${this.deviceId}`
+	}
+
+	get deviceType() {
+		return DeviceType.TELEMETRICS
+	}
+
+	getStatus() {
+		const messages: string[] = []
+
+		switch (this.statusCode) {
+			case StatusCode.GOOD:
+				this.errorMessage = ''
+				messages.push('Connected')
+				break
+			case StatusCode.BAD:
+				messages.push('No connection')
+				break
+			case StatusCode.UNKNOWN:
+				this.errorMessage = ''
+				messages.push('Not initialized')
+				break
+		}
+
+		if (this.errorMessage) {
+			messages.push(this.errorMessage)
+		}
+
+		const deviceStatus: DeviceStatus = {
+			statusCode: this.statusCode,
+			messages,
+			active: this.isActive,
+		}
+		return deviceStatus
+	}
+
+	handleState(newState: TimelineState, mappings: Mappings): void {
+		super.onHandleState(newState, mappings)
+		_.each(newState.layers, (timelineObject, _layerName) => {
+			const telemetricsObject: TimelineObjTelemetrics = timelineObject as unknown as TimelineObjTelemetrics
+			telemetricsObject.content.presetShotIdentifiers.forEach((presetShotIdentifier: number) => {
+				const command = `P0C${presetShotIdentifier}\r`
+
+				if (this.hasCommandBeenSendWithinLastSecond(command, newState.time)) {
+					return
+				}
+
+				this.recentlySendCalls.set(command, newState.time)
+
+				this.doOnTime.queue(newState.time, undefined, () => {
+					this.socket.write(command)
+				})
+			})
+		})
+	}
+
+	private hasCommandBeenSendWithinLastSecond(command: string, time: number): boolean {
+		return this.recentlySendCalls.has(command) && this.recentlySendCalls.get(command)! >= time - 1000
+	}
+
+	async init(options: TelemetricsOptions): Promise<boolean> {
+		return Promise.resolve(this.setupSocket(options.host))
+	}
+
+	private async setupSocket(host: string): Promise<boolean> {
+		this.socket = new Socket()
+
+		this.socket.on('data', (data: Buffer) => {
+			this.emit('debug', `${this.deviceName} received data: ${data.toString()}`)
+		})
+
+		this.socket.on('error', (error: Error) => {
+			this.updateStatus(StatusCode.FATAL, error)
+		})
+
+		this.socket.on('close', (hadError: boolean) => {
+			this.doOnTime.dispose()
```

Comment:
In theory: If the device keeps reconnection then that is because it doesn't have access to a device, so there would be no device to receive the commands anyway?
Or do you think there is an edgecase here that might cause us not to send commands to a connected device?

Reply:
Yes, I believe there is an edge case. I might have been be wrong why that happens, though. I think this line is not the culprit, but it's the fact that `init` never resolves.
For example if you have a timeline consisting of two timeline objects that will alternate infinitely (just for the sake of ease of reproducing this issue with QuickTSR):
```
literal<TimelineObjTelemetrics>({
	id: 't0',
	enable: {
		start: 0,
		duration: 1000,
		repeating: 2000,
	},
	content: {
		deviceType: DeviceType.TELEMETRICS,
		presetShotIdentifiers: [1],
	},
	layer: 'robo',
}),
literal<TimelineObjTelemetrics>({
	id: 't1',
	enable: {
		start: 1000,
		duration: 1000,
		repeating: 2000,
	},
	content: {
		deviceType: DeviceType.TELEMETRICS,
		presetShotIdentifiers: [2],
	},
	layer: 'robo',
}),
```
If the TSR device is started while the TCP server is not running / unavailable (meaning a `close` event will be emitted at least once by the socket), and eventually the server starts up and the connection is established, no commands will be sent.

Reply:
Here's a suggested fix 6a778487588cd62786d564cded2b4d8e2b8becb2

Reply:
Good catch! :) 
I had only tested where I had connected once, then it disconnected and connected again.
It has been updated